---
title: "Translocation Assignment"
author: "Rafael DePillis"
date: "2/22/2022"
output:
  pdf_document: default
  html_document: default
---
  The basic approach of this computational method is to parse through each chromasome in segments of 100,000 nucleotides and count the amount of translocations in each segment. I then calculate the translocation rate per nucleotide of the segment, adding the values of each segment into a masterlist "chromcentiles". I accomplish this quickly and code-efficiently using a double loop. The output of this computation is printing a data frame that is the regions from the masterlist which has a translocation/nucleotide that was at least 1.0 standard deviations above the mean translocation/nucleotide of every segment generated. The value "deviation" is the amount of deviations above the mean and is the indicator of "hotness". This method can be made much more specific by reducing the calue of "parse" which I set to 100,000. Reducing the value gives a higher definition and will identify smaller more dense regions, however it takes exponentially longer and would crash my computer at a value of 10,000. 
  
  Resources used:
https://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf
https://nyu-cdsc.github.io/learningr/assets/data-transformation.pdf

I use these resources a lot when doing any R project just for help with data wrangling. Other than these I would google a few things here and there such as for type casting, but followed no major projects or examples. I also did talk briefly with my statistics professor to verify my method of measuring "hotness".
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r}
if(!require(tidyverse)) install.packages("tidyverse")
library("tidyverse")
```

Installing only necessary package

```{r}
translocationdata <- read_tsv("RF204_Alt055.tlx")
```

Reading in data

```{r}
table(translocationdata['Rname'])
```
```{r}
chr1 <- filter(translocationdata, Rname == 'chr1')
chr2 <- filter(translocationdata, Rname == 'chr2')
chr3 <- filter(translocationdata, Rname == 'chr3')
chr4 <- filter(translocationdata, Rname == 'chr4')
chr5 <- filter(translocationdata, Rname == 'chr5')
chr6 <- filter(translocationdata, Rname == 'chr6')
chr7 <- filter(translocationdata, Rname == 'chr7')
chr8 <- filter(translocationdata, Rname == 'chr8')
chr9 <- filter(translocationdata, Rname == 'chr9')
chr10 <- filter(translocationdata, Rname == 'chr10')
chr11 <- filter(translocationdata, Rname == 'chr11')
chr12 <- filter(translocationdata, Rname == 'chr12')
chr13 <- filter(translocationdata, Rname == 'chr13')
chr14 <- filter(translocationdata, Rname == 'chr14')
chr15 <- filter(translocationdata, Rname == 'chr15')
chr16 <- filter(translocationdata, Rname == 'chr16')
chr17 <- filter(translocationdata, Rname == 'chr17')
chr18 <- filter(translocationdata, Rname == 'chr18')
chr19 <- filter(translocationdata, Rname == 'chr19')
chr20 <- filter(translocationdata, Rname == 'chr20')
chr21 <- filter(translocationdata, Rname == 'chr21')
chr22 <- filter(translocationdata, Rname == 'chr22')
chrM <- filter(translocationdata, Rname == 'chrM')
chrX <- filter(translocationdata, Rname == 'chrX')
chrY <- filter(translocationdata, Rname == 'chrY')
```

Creating sub-data frames so I can loop through a list of them.

Below is the main computation, using a "loopty loop" method to search through all of the chromosomes for regions of statistically high translocation rates in chunks of 100,000. (This variable can be reduced for higher definition).

```{r}
parse <- 100000.0
chromasomes <- list(chr1, chr2, chr3, chr4, chr5, chr6, chr7, chr8, chr9, chr10, chr11, 
chr12, chr13, chr14, chr15, chr16, chr17, chr18, chr19, chr20, chr21, chr22, chrM, chrX, 
chrY)
chromnames <- list("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chr20", "chr21", "chr22", "chrM", "chrX", "chrY")
length <- list(249250621, 243199373, 198022430, 191154276, 180915260, 171115067, 159138663, 146364022, 141213431, 135534747, 135006516, 133851895, 115169878, 107349540, 102531392, 
90354753, 81195210, 78077248, 59128983, 63025520, 48129895, 51304566, 16571, 155270560, 
59373566)

Start <- c()
Amount <- c()
End <- c()
Rname <- c()
translocation_rate <- c()
Size <- c()
counter <- 1

for (chr in chromasomes){
  translocations <- chr$'Junction'
  i <- 0
  while (i < (length[[counter]]/parse)) {
    y = translocations[translocations >= i*parse & translocations < (i*parse)+parse]
    Start <- c(Start, first(y))
    Amount <- c(Amount, length(y))
    End <- c(End, last(y))
    Size <- c(Size, (last(y)-first(y)))
    Rname <- c(Rname, chromnames[[counter]])
    translocation_rate <- c(translocation_rate, length(y)/parse)
    i <- i+1
  }
  
  counter <- (counter + 1)
}
Deviation <- c()
sd <- sd(translocation_rate)
mean <- mean(translocation_rate)
for (num in translocation_rate){
  Deviation <- c(Deviation, (num-mean)/sd)
}
  
chromcentiles <- data.frame(Rname, Start, Amount, End, Size, Deviation)
High_Frequency_Regions <- filter(chromcentiles, Deviation > 1.0)
print(High_Frequency_Regions)
```
Above is the final output of regions with a significantly high rate of translocations per nucleotide. Start and End denote the bounds of the region. Amount denotes the number of translocations found in that region. Rname is the chromasomes on which the region is found. Deviation is the amount of standard deviations above the mean transformations per nucleotide compared to all regions of the genome. 
